File: src/main/python/com/example/tutorial/dsa/medium/matrix/island_counter.py
"""
IslandCounter
----------------------------------
This program counts the number of islands in a 2D grid.
The core problem solved here is Number of Islands (LeetCode #200).

Problem Statement:
    Given an m x n 2D binary grid which represents a map of '1's (land) and '0's (water),
    return the number of islands. An island is surrounded by water and is formed by
    connecting adjacent lands horizontally or vertically.

Real UseCase:
    In a credit card offers system:
    - Identify clusters of related transactions
    - Group connected merchant locations
    - Find connected fraud patterns in transaction matrix

Company Tags: Amazon, Microsoft, Facebook, Google

See: https://leetcode.com/problems/number-of-islands/
"""

import sys
import os
from typing import List
from collections import deque

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "..", ".."))

from com.example.tutorial.common.utils.sample_data_loader import load_offers

# Direction arrays for 4-directional movement (up, down, left, right).
DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def num_islands_dfs(grid: List[List[str]]) -> int:
    """
    Counts islands using DFS (Depth-First Search).

    LOGIC (DFS Flood Fill):
        1. Scan grid cell by cell
        2. When '1' found, increment count and "sink" the entire island
        3. Sinking = DFS to mark all connected land as visited ('0')
        4. Continue scanning for next unvisited island

    Example Walkthrough:
        Grid:          After sinking island 1:
        1 1 0          0 0 0
        1 0 0    ->    0 0 0
        0 0 1          0 0 1

        Found '1' at (0,0), sink it (DFS marks all connected as '0')
        Continue scan, find '1' at (2,2), sink it
        Total islands: 2

    Time Complexity: O(m x n)
        Each cell is visited at most once.
        Like searching a map for countries - once you've colored a country,
        you don't need to visit those cells again.

    Space Complexity: O(m x n)
        Worst case recursion depth if entire grid is one island.
        Can be O(min(m, n)) with iterative BFS approach.

    Args:
        grid: The 2D grid of '1's (land) and '0's (water).

    Returns:
        Number of islands.
    """
    if not grid or not grid[0]:
        return 0

    rows, cols = len(grid), len(grid[0])
    islands = 0

    def dfs(r: int, c: int) -> None:
        """DFS helper to sink an island by marking all connected land as water."""
        # Boundary and water check.
        if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == '0':
            return

        # Mark as visited (sink the land).
        grid[r][c] = '0'

        # Explore all 4 directions.
        for dr, dc in DIRECTIONS:
            dfs(r + dr, c + dc)

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                islands += 1
                dfs(r, c)  # Sink the island.

    return islands


def num_islands_bfs(grid: List[List[str]]) -> int:
    """
    Counts islands using BFS (Breadth-First Search).

    LOGIC (BFS Flood Fill):
        1. Scan grid cell by cell
        2. When '1' found, increment count and BFS to sink entire island
        3. BFS uses queue to visit neighbors level by level
        4. Continue scanning for next unvisited island

    Time Complexity: O(m x n)
        Each cell visited at most once.

    Space Complexity: O(min(m, n))
        Queue size bounded by the smaller dimension.

    Args:
        grid: The 2D grid of '1's (land) and '0's (water).

    Returns:
        Number of islands.
    """
    if not grid or not grid[0]:
        return 0

    rows, cols = len(grid), len(grid[0])
    islands = 0

    def bfs(start_r: int, start_c: int) -> None:
        """BFS helper to sink an island by marking all connected land as water."""
        queue = deque([(start_r, start_c)])
        grid[start_r][start_c] = '0'  # Mark as visited.

        while queue:
            r, c = queue.popleft()

            # Explore all 4 directions.
            for dr, dc in DIRECTIONS:
                nr, nc = r + dr, c + dc

                # Check bounds and if it's land.
                if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == '1':
                    grid[nr][nc] = '0'  # Mark as visited.
                    queue.append((nr, nc))

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] == '1':
                islands += 1
                bfs(r, c)  # Sink the island.

    return islands


def print_grid(grid: List[List[str]]) -> None:
    """Helper method to print a grid."""
    print("Grid:")
    for row in grid:
        print("  " + " ".join(row))


def main():
    """Main function to demonstrate the IslandCounter."""
    print("=== IslandCounter: Connected Components Demo ===\n")

    # Load credit card offers from sample data
    offers = load_offers()
    print(f"Loaded {len(offers)} offers from sample data.\n")

    # Demonstrate island counting
    print("--- Counting Islands (DFS approach) ---\n")

    grid1 = [
        ['1', '1', '1', '1', '0'],
        ['1', '1', '0', '1', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '0', '0', '0']
    ]
    print_grid(grid1)
    print(f"Number of islands (DFS): {num_islands_dfs([row[:] for row in grid1])}\n")

    grid2 = [
        ['1', '1', '0', '0', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '1', '0', '0'],
        ['0', '0', '0', '1', '1']
    ]
    print_grid(grid2)
    print(f"Number of islands (DFS): {num_islands_dfs([row[:] for row in grid2])}\n")

    # BFS approach
    print("--- Counting Islands (BFS approach) ---\n")
    grid3 = [
        ['1', '0', '1', '0', '1'],
        ['0', '1', '0', '1', '0'],
        ['1', '0', '1', '0', '1']
    ]
    print_grid(grid3)
    print(f"Number of islands (BFS): {num_islands_bfs([row[:] for row in grid3])}\n")


if __name__ == "__main__":
    main()

File: src/main/java/com/example/tutorial/dsa/medium/matrix/IslandCounter.java
package com.example.tutorial.dsa.medium.matrix;

import com.example.tutorial.common.datamodel.Offer;
import com.example.tutorial.common.utils.SampleDataLoader;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.stereotype.Component;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * IslandCounter
 * ----------------------------------
 * <p>This program counts the number of islands in a 2D grid.
 * The core problem solved here is Number of Islands (LeetCode #200).
 *
 * <p><b>Problem Statement:</b>
 * Given an m x n 2D binary grid which represents a map of '1's (land) and '0's (water),
 * return the number of islands. An island is surrounded by water and is formed by
 * connecting adjacent lands horizontally or vertically.
 *
 * <p><b>Real UseCase:</b> In a credit card offers system:
 * <ul>
 *   <li>Identify clusters of related transactions</li>
 *   <li>Group connected merchant locations</li>
 *   <li>Find connected fraud patterns in transaction matrix</li>
 * </ul>
 *
 * <p><b>Examples:</b>
 * <ul>
 *   <li>Grid with isolated 1s → Each 1 is an island</li>
 *   <li>Grid with connected 1s → Connected 1s form one island</li>
 * </ul>
 *
 * <p><b>Company Tags:</b> Amazon, Microsoft, Facebook, Google
 *
 * @see <a href="https://leetcode.com/problems/number-of-islands/">LeetCode 200 - Number of Islands</a>
 */
@Component
public class IslandCounter implements CommandLineRunner {
  // Direction arrays for 4-directional movement (up, down, left, right).
  private static final int[] DR = {-1, 1, 0, 0};
  private static final int[] DC = {0, 0, -1, 1};

  public static void main(String[] args) {
    SpringApplication.run(IslandCounter.class, args);
  }

  @Override
  public void run(String... args) throws Exception {
    System.out.println("=== IslandCounter: Connected Components Demo ===\n");

    // Load credit card offers from sample data
    List<Offer> offers = SampleDataLoader.OFFERS_DTO.get();
    System.out.println("Loaded " + offers.size() + " offers from sample data.\n");

    // Demonstrate island counting
    System.out.println("--- Counting Islands (DFS approach) ---\n");

    char[][] grid1 = {
        {'1', '1', '1', '1', '0'},
        {'1', '1', '0', '1', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '0', '0', '0'}
    };
    printGrid(grid1);
    System.out.println("Number of islands (DFS): " + numIslandsDFS(grid1) + "\n");

    char[][] grid2 = {
        {'1', '1', '0', '0', '0'},
        {'1', '1', '0', '0', '0'},
        {'0', '0', '1', '0', '0'},
        {'0', '0', '0', '1', '1'}
    };
    printGrid(grid2);
    System.out.println("Number of islands (DFS): " + numIslandsDFS(grid2) + "\n");

    // BFS approach
    System.out.println("--- Counting Islands (BFS approach) ---\n");
    char[][] grid3 = {
        {'1', '0', '1', '0', '1'},
        {'0', '1', '0', '1', '0'},
        {'1', '0', '1', '0', '1'}
    };
    printGrid(grid3);
    System.out.println("Number of islands (BFS): " + numIslandsBFS(grid3) + "\n");
  }

  /**
   * Counts islands using DFS (Depth-First Search).
   *
   * <p><b>LOGIC (DFS Flood Fill):</b>
   * <ol>
   *   <li>Scan grid cell by cell</li>
   *   <li>When '1' found, increment count and "sink" the entire island</li>
   *   <li>Sinking = DFS to mark all connected land as visited ('0')</li>
   *   <li>Continue scanning for next unvisited island</li>
   * </ol>
   *
   * <p><b>Example Walkthrough:</b>
   * <pre>
   * Grid:          After sinking island 1:
   * 1 1 0          0 0 0
   * 1 0 0    →     0 0 0
   * 0 0 1          0 0 1
   *
   * Found '1' at (0,0), sink it (DFS marks all connected as '0')
   * Continue scan, find '1' at (2,2), sink it
   * Total islands: 2
   * </pre>
   *
   * <p><b>Time Complexity: O(m × n)</b>
   * <br>Each cell is visited at most once.
   * <br><i>Like searching a map for countries - once you've colored a country,
   * you don't need to visit those cells again.</i>
   *
   * <p><b>Space Complexity: O(m × n)</b>
   * <br>Worst case recursion depth if entire grid is one island.
   * <br>Can be O(min(m, n)) with iterative BFS approach.
   *
   * @param grid the 2D grid of '1's (land) and '0's (water)
   * @return number of islands
   */
  public static int numIslandsDFS(char[][] grid) {
    if (grid == null || grid.length == 0) {
      return 0;
    }

    int rows = grid.length;
    int cols = grid[0].length;
    int islands = 0;

    for (int r = 0; r < rows; r++) {
      for (int c = 0; c < cols; c++) {
        if (grid[r][c] == '1') {
          islands++;
          dfs(grid, r, c);  // Sink the island.
        }
      }
    }

    return islands;
  }

  /**
   * DFS helper to sink an island by marking all connected land as water.
   */
  private static void dfs(char[][] grid, int r, int c) {
    int rows = grid.length;
    int cols = grid[0].length;

    // Boundary and water check.
    if (r < 0 || r >= rows || c < 0 || c >= cols || grid[r][c] == '0') {
      return;
    }

    // Mark as visited (sink the land).
    grid[r][c] = '0';

    // Explore all 4 directions.
    for (int d = 0; d < 4; d++) {
      dfs(grid, r + DR[d], c + DC[d]);
    }
  }

  /**
   * Counts islands using BFS (Breadth-First Search).
   *
   * <p><b>LOGIC (BFS Flood Fill):</b>
   * <ol>
   *   <li>Scan grid cell by cell</li>
   *   <li>When '1' found, increment count and BFS to sink entire island</li>
   *   <li>BFS uses queue to visit neighbors level by level</li>
   *   <li>Continue scanning for next unvisited island</li>
   * </ol>
   *
   * <p><b>Time Complexity: O(m × n)</b>
   * <br>Each cell visited at most once.
   *
   * <p><b>Space Complexity: O(min(m, n))</b>
   * <br>Queue size bounded by the smaller dimension.
   *
   * @param grid the 2D grid of '1's (land) and '0's (water)
   * @return number of islands
   */
  public static int numIslandsBFS(char[][] grid) {
    if (grid == null || grid.length == 0) {
      return 0;
    }

    int rows = grid.length;
    int cols = grid[0].length;
    int islands = 0;

    for (int r = 0; r < rows; r++) {
      for (int c = 0; c < cols; c++) {
        if (grid[r][c] == '1') {
          islands++;
          bfs(grid, r, c);  // Sink the island.
        }
      }
    }

    return islands;
  }

  /**
   * BFS helper to sink an island by marking all connected land as water.
   */
  private static void bfs(char[][] grid, int startR, int startC) {
    int rows = grid.length;
    int cols = grid[0].length;

    Queue<int[]> queue = new LinkedList<>();
    queue.offer(new int[] {startR, startC});
    grid[startR][startC] = '0';  // Mark as visited.

    while (!queue.isEmpty()) {
      int[] cell = queue.poll();
      int r = cell[0];
      int c = cell[1];

      // Explore all 4 directions.
      for (int d = 0; d < 4; d++) {
        int nr = r + DR[d];
        int nc = c + DC[d];

        // Check bounds and if it's land.
        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && grid[nr][nc] == '1') {
          grid[nr][nc] = '0';  // Mark as visited.
          queue.offer(new int[] {nr, nc});
        }
      }
    }
  }

  /**
   * Helper method to print a grid.
   */
  private static void printGrid(char[][] grid) {
    System.out.println("Grid:");
    for (char[] row : grid) {
      System.out.print("  ");
      for (char cell : row) {
        System.out.print(cell + " ");
      }
      System.out.println();
    }
  }
}
